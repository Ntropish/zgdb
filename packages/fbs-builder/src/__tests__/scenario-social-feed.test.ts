import { describe, it, expect } from "vitest";
import {
  createFbsBuilder,
  createInitialFbsFileState,
  renderFbs,
} from "../index.js";

describe("Scenario Test: EchoSphere - Social Media Feed Item", () => {
  it("should generate a schema for a polymorphic social media feed", async () => {
    const builder = createFbsBuilder();
    const initialState = createInitialFbsFileState();

    /**
     * SCENARIO: The "EchoSphere" social media platform needs to represent
     * items in a user's feed. A feed item is a container that holds metadata
     * (like post ID, author) and the content itself, which can be a simple
     * text message, a photo gallery, or an interactive poll.
     *
     * FEATURES TESTED:
     * - A complex `union` to handle different types of content.
     * - A `struct` for simple, related data (`PhotoMetadata`).
     * - Multiple tables, some of which are only used within the union.
     * - `deprecated` attribute on a field to test schema evolution.
     * - Vectors of strings for poll options and photo URLs.
     */

    builder.namespace("EchoSphere.Feed");

    // A simple text post.
    builder
      .table("TextContent")
      .docs("Content for a simple text-based post.")
      .field("text", "string");

    // A struct for photo metadata.
    builder
      .struct("PhotoMetadata")
      .docs("Metadata for a single photo.")
      .field("width", "ushort")
      .field("height", "ushort");

    // A gallery of photos.
    builder
      .table("GalleryContent")
      .docs("Content for a gallery post with multiple photos.")
      .field("urls", "string", { isVector: true })
      .field("metadata", "PhotoMetadata", { isVector: true });

    // An interactive poll.
    builder
      .table("PollContent")
      .docs("Content for an interactive poll.")
      .field("question", "string")
      .field("options", "string", { isVector: true });

    // The union that allows a FeedItem to contain any of the content types.
    builder
      .union("Content")
      .docs("A union of all possible content types for a feed item.")
      .value("TextContent")
      .value("GalleryContent")
      .value("PollContent");

    // The main container for any item appearing in the feed.
    builder
      .table("FeedItem")
      .docs("The root container for an item on the social media feed.")
      .field("id", "ulong", { attributes: { key: true } })
      .field("author_id", "string")
      .field("likes", "uint", { defaultValue: 0 })
      .field("old_thread_id", "string", { attributes: { deprecated: true } })
      .field("content", "Content", { attributes: { required: true } });

    builder.root_type("FeedItem");

    // Build and render the schema
    const finalState = await builder.build(initialState);
    const result = renderFbs(finalState);

    // The output will be compared against a snapshot.
    expect(result).toMatchSnapshot();
  });
});
