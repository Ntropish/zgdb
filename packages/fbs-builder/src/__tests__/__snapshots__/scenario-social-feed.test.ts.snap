// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Scenario Test: EchoSphere - Social Media Feed Item > should generate a schema for a polymorphic social media feed 1`] = `
"namespace EchoSphere.Feed;
/// Content for a simple text-based post.
table TextContent {
  text: string;
}

/// Metadata for a single photo.
struct PhotoMetadata {
  width: ushort;
  height: ushort;
}

/// Content for a gallery post with multiple photos.
table GalleryContent {
  urls: [string];
  metadata: [PhotoMetadata];
}

/// Content for an interactive poll.
table PollContent {
  question: string;
  options: [string];
}

/// A union of all possible content types for a feed item.
union Content {
  TextContent,
  GalleryContent,
  PollContent,
}

/// The root container for an item on the social media feed.
table FeedItem {
  id: ulong (key);
  author_id: string;
  likes: uint = 0;
  old_thread_id: string (deprecated);
  content: Content (required);
}

root_type FeedItem;"
`;
